File: backend-go/services/whatsapp.go
package services

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

func SendMessage(to, text string) error {
	phoneNumberID := os.Getenv("WHATSAPP_PHONE_NUMBER_ID")
	accessToken := os.Getenv("WHATSAPP_ACCESS_TOKEN")

	url := fmt.Sprintf("https://graph.facebook.com/v19.0/%s/messages", phoneNumberID)

	payload := map[string]any{
		"messaging_product": "whatsapp",
		"to":                to,
		"type":              "text",
		"text":              map[string]string{"body": text},
	}

	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+accessToken)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("WhatsApp API error (%d): %s", resp.StatusCode, string(b))
	}
	return nil
}

func ResolveAudioURL(mediaID string) (string, error) {
	accessToken := os.Getenv("WHATSAPP_ACCESS_TOKEN")
	url := fmt.Sprintf("https://graph.facebook.com/v19.0/%s", mediaID)

	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", "Bearer "+accessToken)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	audioURL, _ := result["url"].(string)
	return audioURL, nil
}

File: whatsapp.js
const WHATSAPP_API_URL = `https://graph.facebook.com/v19.0/1050575971469681/messages`;

export async function sendWhatsAppMessage(to, text) {
  const payload = {
    messaging_product: "whatsapp",
    to,
    type: "text",
    text: { body: text },
  };

  const response = await fetch(WHATSAPP_API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer EAAU6rMEJ704BRLFS4zVpTEd48qTx4hubAZBaKNIXkpDJ0DDoHZCRZBQrWlA3ldiDCzXO2C59GrwNgDoXgSr9JaOvLbZAPfd4kwQ4C7ZCS7WKsOJji1UT6tdDYJ0gCGfu8X4PicnIN6UjG4vh7LJcPwG6OBeqs8Q8duidZADLufuZCJibHNwRysazHxeT3Ts0W8ZAxZBPX4L4X7qWbxFEOJTr9DRn3ZAR0cbTR1brioGpuZCSN7C81JUpsrgPxfV5ZAHqiXHe7E9Rq1mS7qpDVHNVz9VFZBi4d`,
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`WhatsApp API error: ${JSON.stringify(error)}`);
  }

  return response.json();
}